plugins/catalog-backend/src/processing/util.ts

```diff
@@ -86,3 +86,50 @@ export function isObject(value: JsonValue | undefined): value is JsonObject {
 export const validateEntity = entitySchemaValidator();
 
 export const validateEntityEnvelope = entityEnvelopeSchemaValidator();
+
+/**
+ * Creates a barrier with a timeout, that can be awaited or prematurely
+ * released either manually or by an abort signal.
+ */
+export function createBarrier(options: {
+  waitTimeoutMillis: number;
+  signal: AbortSignal;
+}): {
+  wait: () => Promise<void>;
+  release: () => void;
+} {
+  const { waitTimeoutMillis, signal } = options;
+  const resolvers: Array<() => void> = [];
```

Comment:
Bit simpler to make this a `Set` I'd think

Reply:
Agreed! Did that now